### PR TITLE
ARROW-3254: [C++] Add SOURCES option to ADD_ARROW_TEST to be able to combine multiple test files into a single executable

### PR DESCRIPTION
**Pull request by wesm**
I also moved arrow/test-util.h to arrow/util/testing.h and moved some of the implementation of that file into arrow/util/testing.cc, so that many of these auxiliary test functions are only compiled once.

### CODE REVIEW

File: cpp/cmake_modules/BuildUtils.cmake [hunk 1]
```diff
@@ -397,40 +398,44 @@ function(ADD_ARROW_TEST REL_TEST_NAME)
...
+  if (ARG_EXTRA_LINK_LIBS)
+    target_link_libraries(${TEST_NAME} ${ARG_EXTRA_LINK_LIBS})
+  endif()
 
-    foreach (TEST_LABEL ${ARG_LABELS})
-      add_dependencies(${TEST_LABEL} ${TEST_NAME})
-    endforeach()
-  else()
-    # No executable, just invoke the test (probably a script) directly.
-    set(TEST_PATH ${CMAKE_CURRENT_SOURCE_DIR}/${REL_TEST_NAME})
```
Comment (wesm) [inline cpp/cmake_modules/BuildUtils.cmake]: This was unused

**Comment by wesm** [comment]
Argh Windows is a constant source of problems. @pitrou apparently if you combine a DLL (like arrow.dll) with a static library (like the auxiliary testing static library I created here) it will gripe about duplicated inline symbols. I think we should revert to statically linking the tests on Windows to avoid this madness

**Comment by pitrou** [comment]
I'd like to step back a bit:
* having multiple test files allow parallelizing (an important feature when developing, especially now that Parquet tests have been added to the bunch - some of them are quite slow)
* linking should be fast if Arrow is linked as a shared library
* the test executables should be small if Arrow is linked as a shared library

The fact that we do/did static linking for tests is an anomaly, and is IMHO the real source of the slowness and disk consumption.

**Comment by wesm** [comment]
It appears that the issue is that the symbols in the DLL have `dllimport` linkage while in the static lib they have dllexport: https://akrzemi1.wordpress.com/2016/11/28/the-one-definition-rule/. I'm going to fire up my windows box and see if I can fix it

**Comment by wesm** [comment]
> linking should be fast if Arrow is linked as a shared library

My understanding is that code coverage doesn't work if you don't statically link

**Comment by wesm** [comment]
Reading https://github.com/apache/kudu/blob/master/CMakeLists.txt#L422, it seems that maybe the coverage issue goes away if you use Clang

**Comment by pitrou** [comment]
Do you mean codecov-generated code coverage results?

**Comment by wesm** [comment]
Yeah, apparently there is an issue with code coverage data generated when compiling with `--coverage` and using gcc. Read this commit message for more information about this https://github.com/apache/kudu/commit/48799d3fed53c79a5163fd9c4cd9cbae80a9f2d6

**Comment by wesm** [comment]
I'll revisit this after https://github.com/apache/arrow/pull/2997 is merged

**Pull request closed without merging**